### DISCUSSION TITLE
Encounter Self-referential type when using winit(>= 0.30) `Window` and wgpu's `Surface`

### DISCUSSION THREAD

**Discussion by ifsheldon** [gfx-rs/wgpu#6005] [Q&A] [answered]
Hi all! I am maintaining a Direct Volume Rendering(DVR) app which depends on `wgpu` and `winit`. When I tried to update my app to use `winit` >= 0.30, I was stuck in the problem of self reference. Do you have any ideas to resolve this?

A bit of background: 

* Recently, `winit` decided to make their event handling APIs to be trait-based. In its documentation, the example needs to store a `window: Option<Windows>` for `resume` method and requesting redraws. 
  https://github.com/rust-windowing/winit/blob/73c01fff96266a6a7d9159e9bd26acc00df0bbc9/src/lib.rs#L79
* A `Surface` has a reference to a window, in the case of my app, to `winit::Window`
  https://docs.rs/wgpu/22.0.0/src/wgpu/lib.rs.html#584-606

In my case, I need these in my App states:
* `window: winit::Window` to request redraw and get its size if the window's scaling factor gets changed.
* `surface: wgpu: Surface<'window>` to use `Surface::get_current_texture` before doing the actual rendering.

I can use `ouroboros` to create self-referential types but I don't know what will happen if a `Window` got replaced by another during `winit::application::ApplicationHandler::resumed`.

So I have actually 3 questions, which may be issues when using wgpu and winit in combo:
1. How can I store `winit::Window` and `wgpu::Surface<'window>` together?
2. Is there a potential dangling pointer problem when a `winit::Window` gets replaced during `winit::application::ApplicationHandler::resumed`?
3. How can we avoid this dangling pointer problem if it may exist?

Further references:
* My code is opensourced on https://github.com/nanovis/Wenderer
   * the `main` branch works with `wgpu 22.0.0` and `winit 0.29`
   * the `winit0.30` branch fails compilation due to self reference.....

**Comment by cwfitzgerald** [comment, marked as answer]
1. You can put the window in an Arc, then hand the surface a clone of the Arc. The lifetime will be static.
2. Because of the Arc, everything should work fine

**Reply by ifsheldon** [reply]
Awesome! It takes a bit refactoring but it works! Thanks a lot!
